clr/commands.py

```diff
@@ -1,39 +1,47 @@
 from __future__ import print_function
 from importlib import import_module
-from builtins import zip
-from builtins import object
 from dataclasses import dataclass
 import inspect
+from inspect import Signature
 import sys
 import textwrap
 import types
 import difflib
+import shelve
+import os
+from collections import namedtuple
 
 import clr.config
 
 # Sentinal for args get in get_command_spec to indicate there is no default.
-NO_DEFAULT = object()
+# Because we are pickling command specs for clr cache use a random int
+# and check for equality rather than object identity.
+# TODO(michael.cusack): Move command spec to inspect.Signature and remove this.
+NO_DEFAULT = 4194921784511160246
+
+# Sorted list of command namespace keys.
+NAMESPACE_KEYS = sorted(clr.config.commands().keys() | {'system'})
 
-NAMESPACE_KEYS = clr.config.commands().keys() | {'system'}
 # Load lazily namespace modules as needed. Some have expensive/occasionally
 # failing initialization.
 __namespaces = {}
 
-def get_namespaces():
-    # Fill namespace cache
-    for key in NAMESPACE_KEYS: get_namespace(key)
-    return __namespaces
-
 def _load_namespace(key):
+    """Imports the module specified by the given key."""
     if key == 'system':
-        return System()
-
-    module_path = clr.config.commands()[key]
-    try:
-        module = import_module(module_path)
-        return module.COMMANDS
-    except Exception as e:
-        return ErrorLoadingNamespace(key, e)
+        instance = System()
+    else:
+        module_path = clr.config.commands()[key]
+        try:
+            module = import_module(module_path)
+            instance = module.COMMANDS
+        except Exception as e:
+            return ErrorLoadingNamespace(key, e)
+    descr = instance.descr
+    longdescr = instance.longdescr if hasattr(instance, 'longdescr') else descr
+    command_callables = {a[4:]: getattr(instance, a) for a in dir(instance) if a.startswith('cmd_')}
+    command_specs = {n: get_command_spec(c) for n, c in command_callables.items()}
```

Comment:
Can we avoid the one-character variables and use something a bit more descriptive for clarity? Hopefully too many people aren't poking in the guts of this method, but it's still nice to keep it as readable as possible!

Reply:
Sure. I usually only use them when it's inline and obvious.

`[t.foo for t in things]`

These got...a little complicated.